server/app/services/program/predicate/PredicateGenerator.java

```diff
@@ -0,0 +1,290 @@
+   * @param roQuestionService a {@link ReadOnlyQuestionService} for validating that questions
+   *     referenced in the form are active or draft.
+   * @throws QuestionNotFoundException if the form references a question ID that is not in the
+   *     {@link ReadOnlyQuestionService}.
+   * @throws BadRequestException if the form is invalid.
+   */
+  public PredicateDefinition generatePredicateDefinition(
+      DynamicForm predicateForm, ReadOnlyQuestionService roQuestionService)
+      throws QuestionNotFoundException {
+    Multimap<Integer, LeafOperationExpressionNode> leafNodes = LinkedHashMultimap.create();
+
+    HashSet<String> consumedMultiValueKeys = new HashSet<>();
+    final PredicateAction predicateAction;
+
+    try {
+      predicateAction = PredicateAction.valueOf(predicateForm.get("predicateAction"));
+    } catch (IllegalArgumentException e) {
+      throw new BadRequestException(
+          String.format(
+              "Missing or unknown predicateAction: %s", predicateForm.get("predicateAction")));
+    }
+
+    for (String key : predicateForm.rawData().keySet()) {
+      Matcher singleValueMatcher = SINGLE_PREDICATE_VALUE_FORM_KEY_PATTERN.matcher(key);
+      Matcher multiValueMatcher = MULTI_PREDICATE_VALUE_FORM_KEY_PATTERN.matcher(key);
+
+      int groupId;
+      long questionId;
+      Scalar scalar;
+      Operator operator;
+      PredicateValue predicateValue;
+
+      if (singleValueMatcher.find()) {
+        Pair<Integer, Long> groupIdQuestionIdPair =
+            getGroupIdAndQuestionId(roQuestionService, singleValueMatcher);
+        groupId = groupIdQuestionIdPair.getLeft();
+        questionId = groupIdQuestionIdPair.getRight();
+        Pair<Scalar, Operator> scalarOperatorPair = getScalarAndOperator(predicateForm, questionId);
+        scalar = scalarOperatorPair.getLeft();
+        operator = scalarOperatorPair.getRight();
+
+        predicateValue =
+            parsePredicateValue(scalar, operator, predicateForm.get(key), ImmutableList.of());
+      } else if (multiValueMatcher.find() && !consumedMultiValueKeys.contains(key)) {
+        Pair<Integer, Long> groupIdQuestionIdPair =
+            getGroupIdAndQuestionId(roQuestionService, multiValueMatcher);
+        groupId = groupIdQuestionIdPair.getLeft();
+        questionId = groupIdQuestionIdPair.getRight();
+        Pair<Scalar, Operator> scalarOperatorPair = getScalarAndOperator(predicateForm, questionId);
+        scalar = scalarOperatorPair.getLeft();
+        operator = scalarOperatorPair.getRight();
+
+        ImmutableList<String> multiSelectKeys =
+            predicateForm.rawData().keySet().stream()
+                .filter(
+                    filteredKey ->
+                        filteredKey.startsWith(
+                            String.format(
+                                "group-%d-question-%d-predicateValues", groupId, questionId)))
+                .collect(ImmutableList.toImmutableList());
```

Comment:
you can replace this and the next line with 

`.forEach(consumedMultiValueKeys::add);`

Reply:
`multiSelectKeys` is referenced again on L118:
```java
        ImmutableList<String> rawPredicateValues =
            multiSelectKeys.stream()
                .map(predicateForm.rawData()::get)
                .collect(ImmutableList.toImmutableList());
```